test: Fix pod cleanup after various tests

joestringer opened the pull request:
See individual commits for more details.

According to the [Cilium Ginkgo e2e test docs](https://docs.cilium.io/en/latest/contributing/testing/e2e/#example-test-layout), All `AfterEach` statements are executed before all `AfterAll` statements.

Previously, some of this code was assuming that `AfterAll` inside a context would run first (to delete pods) and then the `AfterEach` would run outside the context (to wait for the pods to terminate), then `AfterAll` outside the context to finally clean up the Cilium pods. The result was that in issue #18447, the next test to run could hit a race condition where pods were deleted but did not fully terminate before Cilium was removed. Cilium ends up getting deleted before all the pods, which means that there is no longer any way to execute the `CNI DEL` operation for those pods, so they get stuck in `Terminating` state.

Other parts of the code seemed to just omit the check for whether the pods were correctly terminated or not.

Fix these isseus by moving checks that the test pods are fully terminated into the statements where the pods are deleted.

Fixes: 412e299df74f ("test: Move LRP tests to a separate suite")
Fixes: #18447
Fixes: #18566

Note, I initially marked this for backport to v1.10 based on the first LRP test refactor commit, but left it like this since it seems like it should make the testsuite a bit more robust in general. I don't know if each of the changes will successfully backport to v1.10 or not. If not, we can maybe just drop those changes from the commit. Or we could just decide to avoid backporting the PR to v1.10 altogether.

joestringer committed:
test: Fix pod cleanup after LRP test group

According to the Ginkgo e2e test docs, All "AfterEach" statements are
executed before all "AfterAll" statements[1].

Previously, this code was assuming that AfterAll inside a context would
run first (to delete pods) and then the AfterEach would run outside the
context (to wait for the pods to terminate), then AfterAll outside the
context to finally clean up the Cilium pods.

The result was that in issue 18447, the next test to run could
hit a race condition where pods were deleted but did not fully terminate
before Cilium was removed. Cilium ends up getting deleted before all the
pods, which means that there is no longer any way to execute the CNI DEL
operation for those pods, so they get stuck in Terminating state.

Fix it by moving the check that the test pods are fully terminated
before returning from the AfterAll statement where we initiate the
deletion of those pods.

[0]: https://docs.cilium.io/en/latest/contributing/testing/e2e/#example-test-layout

Signed-off-by: Joe Stringer <joe@cilium.io>

nbusseneau reviewed (approved):
Well spotted!

joestringer committed:
test: Fix external_ips test cleanup

This test seems like it's always skipped, so who knows whether it works
or not. But this should at least mitigate issues with cleaning up state
at the end of the test, similar to the issues in the adjacent commits.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
test: Clean up pods after deleting them

The library in the tree that we use to interact with kubectl doesn't
automatically wait for all pods to be deleted when we delete them.
Depending on which test runs next, this can mean that we occasionally
hit a race condition like the one in Issue 18447:

* Delete Pods
* Delete Cilium
* Pods can't be fully deleted because Cilium is down
* BeforeAll for next test waits to make sure everything is ready so that
  we can install Cilium
* The next test times out while attempting this process

This commit goes through all of the existing suites, removes cases where
we are relying on incorrect scheduling between AfterEach/AfterAll
statements, and adds in waits after deleting pods anywhere in the suite
where we delete pods.

Signed-off-by: Joe Stringer <joe@cilium.io>

gandro commented on test/helpers/manifest.go:
```diff
@@ -280,6 +280,8 @@ func (m *DeploymentManager) DeleteAll() {
 // DeployCilium()
 func (m *DeploymentManager) DeleteCilium() {
 	if m.ciliumDeployed {
+		/* Ensure any Cilium-managed pods are terminated first */
```
tiny nit:
```suggestion
		// Ensure any Cilium-managed pods are terminated first
```

gandro reviewed (approved):
Thanks a ton for this!

aditighag reviewed (approved):
Good find! Thanks!

brb reviewed (approved):
Nice find and thanks for the fix!

nbusseneau reviewed (approved):
Changes LGTM.

joestringer committed:
test: Improve error reporting when wait fails

When the test fails because it is waiting for containers to terminate,
add this information into the error that is reported so that developers
can more easily debug the issue that's occurring.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
test: Properly skip K8sBandwidth tests

Previously the skip was on the Context which meant that the
BeforeAll/AfterAll for the K8sBandwidthTest Describe would be run, which
could deploy pods and fail to clean them up properly. Skip these
unnecessary steps as they can cause problems in test runs where this
test is skipped, and because it's unnecessary work.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
test: Properly skip CustomCalls test

Previously the skip was on the Context which meant that the
BeforeAll/AfterAll for the K8sCustomCalls Describe would be run, which
could deploy pods and fail to clean them up properly. Skip these
unnecessary steps as they can cause problems in test runs where this
test is skipped, and because it's unnecessary work.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
test: Fix CustomCalls pod teardown

Deleting all pods after deleting Cilium is a sure-fire way to introduce
flakiness where pods hang in Terminating state during deletion, because
the Cilium pods have been cleaned up and kubelet can't call into the CNI
to properly clean up the pods.

Delete the pods first and wait for termination to finish.

CC: Quentin Monnet <quentin@isovalent.com>
Fixes: 9d4e99d500d0 ("tests: rework custom calls's AfterEach/AfterAll blocks to skip if needed")
Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
helpers: Scale DNS pods up/down with replicas

Use the spec.replicas field in the kube-dns deployment to reconfigure
the number of DNS pods in the cluster during pod restart. This should
restart the kube-dns pods correctly, similar to just deleting the pods.

A subsequent patch intends to split this usage out depending the status
of Cilium deployments, in order to mitigate issues where kube-dns pods
are deployed differently between test runs due to Cilium configuration
(like endpoint routes mode).

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
test: Clean up DNS pods more gracefully

As of the writing of this patch, Cilium does not support transitioning
between enable-endpoint-routes=true and enable-endpoint-routes=false.

Commit c18cfc874620 ("test: Redeploy DNS pods in AfterAll for datapath tests")
attempted to ensure that DNS pods are properly deleted from nodes prior
to completing a particular test file. This was to ensure that when
making changes to Cilium configuration like toggling endpoint routes
mode, the pods are properly switched over to the new mode.

Previously, the order of the test cleanup could cause problems with
terminating the kube-dns pods because Cilium was first deleted from the
cluster, but the kube-dns pods would get stuck in "Terminating" state.
This is believed to be because the nodes would still attempt to call the
cilium-cni binary in order to remove the DNS pods, and that would
attempt to reach out to the cilium-agent on the node, which is no longer
present.

The motivation behind this patch, then, is to first scale the pods down
to zero so that no kube-dns pods are present in the cluster during test
file cleanup, then shut down Cilium. Finally, we restore the number of
replicas back to the original value, so that when the next test sets
Cilium back up, the kubedns pods will be ready to be deployed afresh.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
test: Clean up after each graceful termination test

Previously this code was mismatched: The BeforeEach would attempt to
re-create the graceful termination app before each of the It statements,
however the cleanup would only occur once.

While preparing PR 18448, I hit an issue GH-18634 where if we didn't
wait sufficiently long between individual tests in the graceful
termination test Context, the test would fail during the following
BeforeEach because the previous test's client was not yet cleaned up.

Fix this by ensuring the cleanup & wait for pod termination occurs in an
AfterEach for each It statement in this Context, to match the behaviour
of the overall test. The next commit will then shift the cleanup logic
out of a Describe level AfterEach and into the individual Contexts.

Signed-off-by: Joe Stringer <joe@cilium.io>

aditighag reviewed (commented):
Reviewed d4739ad757d8d84dd65302e0e85e985b51198d69. The change looks correct to me.

Thanks! 🙏 

While at it, we should remove this duplicate code that deploys the yaml - https://github.com/cilium/cilium/blob/master/test/k8sT/Services.go#L1639-L1641. I can fix this in a separate PR if you don't want to do it in this PR. 

>The next commit will then shift the cleanup logic
out of a Describe level AfterEach and into the individual Contexts.

Can you confirm the commit? Might as well double check that too.
